[Step 3] 연락처관리 UI ― iyeah, Jenna

ueunli opened the pull request:
안녕하세요 Cory !! @corykim0829
주말 잘 보내셨나요? 🐶
저희는 부지런히 Step을 진행하여 조금은 이르게 PR을 보내게 되었습니다!
전반적으로 저번 프로젝트와 비슷한 부분이 많지만 새롭게 고민한 포인트도 많았던 것 같아요 ㅎㅎ
놓친 부분이 있을까 싶은데, 조언 주실 부분이 있다면 가감없이 말씀해 주세요
월요일인데 이번 주도 파이팅 하시고 리뷰 잘 부탁드립니다 🤍


# ✅ 구현 사항
- [x]  1️⃣ 화면 간 데이터 전달 - 델리게이트
- [x]  2️⃣ 텍스트필드 값을 입력양식에 맞춰 실시간 변환하기

![](https://i.imgur.com/XZka8tF.png)
![](https://i.imgur.com/osAnaRP.png)

   https://user-images.githubusercontent.com/67406889/216862939-2b08f03d-490b-49d1-a949-2d2418f65c52.mp4

(↑ 테스트 목적으로 디버깅영역에 `print(textField.text!)`한 결과입니다. 커밋내역에는 포함되지 않았습니다!)

　
# ✅ 구현 과정
## 1️⃣　화면 이동 시 델리게이트 패턴으로 데이터 전달
> 새 연락처를 입력하고 Save 버튼을 누르면 연락처를 저장하고 연락처 목록에 띄우기

이름, 나이, 연락처를 작성하고 첫번째 화면에 데이터를 어떻게 전달해야 할까 고민하다가 델리게이트 패턴을 적용하였어요.
`ListProfileViewController`가 `ListProfileViewControllerDelegate` 프로토콜 채택
* `AddProfileViewController` 텍스트 필드에서 입력받은 이름, 나이, 연락처를 전달하고
* `ListProfileViewController` 에서 전달받은 데이터로 `updateProfile`에서 아래와 같은 동작 실행
    * `Profile` 생성
    * `contactManageSystem.addProfile(of:)` 로 연락처 목록에 추가 
    * 테이블 뷰 `reloadData()` 실행

　
## 2️⃣　텍스트필드 값변경
> 아이폰 기본 전화앱에서처럼 입력한 자릿수에 따라 하이픈(-)을 추가

`.phonePad`는 `-`입력을 지원하지 않아, 텍스트필드에 값이 입력될 때마다 `-`가 적절한 위치에 삽입되게끔 자동변환 해주는 로직을 구현했습니다.
- `AddProfileViewController`가 `UITextFieldDelegate프로토콜을` 채택
- `textField(_:shouldChangeCharactersIn:replacementString:)`메서드로 새로운 입력의 종류에 따라 연락처 텍스트필드에서의 입력을 허용하거나, 또는 양식에 맞춘 변환값으로 대체하여 직접 할당(입력 거절)
- `PhoneNumberRegularExpressions`열거형으로 자릿수별 변환 방식을 정의

　
　

ueunli committed:
<feat> 스토리보드 내 수정사항

- 기존 뷰에서 모달 방식으로 전환할 새로운 뷰 추가
- 각 뷰에 navigation controller와 bar button item 추가

iyeahh committed:
<feat> 새 연락처 view의 세팅

- 이름, 나이, 연락처 Label과 TextFiled 추가
- 스택뷰를 이용하여 오토레이아웃 설정

ueunli committed:
<feat> AddProfileViewController 추가

- 네이밍 변경) ViewController → ListProfileViewController
- 추가) AddProfileViewController

iyeahh committed:
<feat> 텍스트필드 연결 및 키보드 설정

ueunli committed:
<feat> 두 뷰의 화면전환 방식 변경 및 델리게이트 구현

- AddProfileViewControllerDelegate 추가
- 화면전환 방법 구현 위치를 스토리보드 → 코드로 변경

iyeahh committed:
<feat> AddProfileView의 navigation bar button 코드 연결

- saveButtonDidTap
- cancelButtonDidTap 얼럿 생성

ueunli committed:
<feat> saveButtonDidTap 얼럿 생성

- InputError의 원시값 및 계산속성 `message` 정의
- isValidUserInput 메서드 리팩토링

iyeahh committed:
<refactor> Step2 피드백 반영

- dummy data를 VC의 상수로 할당
- 프로필 목록을 테이블 뷰에 표시하기 전에 오름차순으로 정렬하는 연산속성 추가

ueunli committed:
<feat> 프로필 업데이트 방식 변경

- 델리게이트 프로토콜 수정
- 전달받은 정보를 tableView에 반영
- 기존 `addProfile()`를 리팩토링

iyeahh committed:
<feat> 연락처를 표준 전화번호 양식에 맞춰 변환하는 메서드 추가

- telTextField의 키보드 타입을 .phonePad로 변경
- 자릿수에 따라 정규표현식대로 변환

ueunli committed:
<refactor> 연락처 정규표현식 변환 함수 가독성 개선

iyeahh committed:
<refactor> alert 생성 메서드 별도로 구현

- save 버튼 누르면 makeAlertToInform(_: )
- cancel 버튼 누르면 makeAlertToAsk()

ueunli committed:
<feat> 이름 띄어쓰기 자동 제거

iyeahh committed:
<conf> 폴더링

ueunli committed:
<feat> 연락처 자동변환 기능 추가

- telTextField의 입력값 자릿수에 따라 하이픈(`-`)을 삽입하는 기능 추가
- 새 프로필 검증을 위한 정규표현식에서도 하이픈(`-`)을 허용하도록 수정

ueunli committed:
<fix> 연락처 정규표현식 수정

- 하이픈(-)을 포함하여 11자리가 되어야 등록되도록 변경

iyeahh committed:
<refactor> 접근제어 변경, 개행 수정

ueunli committed:
<refactor> 세부사항 수정

- 불필요한 delegate 해제
- nameTextField의 키보드 추천·자동완성 해제
- nameTextField에 공백만 입력해도 등록되는 문제 해결 -  에러처리 되도록 검증절차 추가

iyeahh committed:
<feat> 텍스트 필드 외의 영역을 터치 하였을 때 키보드 내리도록 구현

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+        ageTextField.keyboardType = .numberPad
+        telTextField.keyboardType = .phonePad
+        nameTextField.autocorrectionType = .no
+        telTextField.delegate = self
```
TextField delegate 선언이 두번 되어있네요 정리해주세요 :)

ueunli replied:
수정했어요! 감사합니다♥

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+        telTextField.delegate = self
+    }
+    
+    @IBAction private func canelButtonDidTap(_ sender: UIBarButtonItem) {
```
함수명에 오타가 있는 것 같아요 수정해주세요 :)

ueunli replied:
헉 바로 수정했어요!!! 알려주셔서 감사합니다!!!🙆‍♀️

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+    }
+    
+    @IBAction private func canelButtonDidTap(_ sender: UIBarButtonItem) {
+        let alert = makeAlertToAsk()
+        present(alert, animated: true)
```
함수로 UIAlertController 구현부를 분리한 부분은 깔끔해서 좋습니다!
하지만 UIAlertController를 반환하지 않고 해당 함수에서 present까지 처리하면 더 깔끔할 것 같아요!

ueunli replied:
너무 좋은 방법인 것 같아요! 조언 감사합니다 present까지 해주도록 수정했어요💚

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+
+extension AddProfileViewController: UITextFieldDelegate {
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        guard textField.keyboardType == .phonePad else { return true }
```
특정 textField를 처리하기 위해, keyboardType 요소를 사용하기보다는, 해당 인스턴스를 이용하면 좋을 것 같은데 어떻게 생각하시나요?

ueunli replied:
```Swift
guard textField == telTextField else { return true }
```
위와 같이 고쳐봤어요🙂 말씀하신 대로 해당 인스턴스를 이용하는 편이 가독성에도 좋고 훨씬 직관적인 것 같아요 !!

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+            return
+        }
+        
+        let alert = makeAlertToInform(message)
+        present(alert, animated: true)
```
여기도 위에 피드백과 마찬가지로 함수에서 모두 처리되면 좋을 것 같아요 :)

ueunli replied:
`makeAlertToAsk()`함수와 동일하게 수정 완료했어요 :D

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+    }
+    
+    @IBAction func saveButtonDidTap(_ sender: UIBarButtonItem) {
+        guard let message = generateAlertMessage() else {
```
generateAlertMessage만 보면 왜 guard문을 사용해서 예외처리가 되는지 이해가 잘 안되는 것 같아요.
에러라고 명시가 되어야 이해가 잘 될 것 같아요!

그리고 보통 guard문의 else에서 에러를 핸들링을 하기 때문에 수정하면 좋을 것 같아요!

ueunli replied:
generateAlertMessage()는 원래 String?을 반환하였는데 Result 타입을 반환하도록 변경하였어요!
### 고민 포인트
네이밍부터 generateErrorMessage()로 수정한 뒤 말씀하신 부분에 대해 천천히 의논해 봤어요 :)
에러가 발생하지 '않으면' nil을 반환하는 메서드였기에, guard문의 로직을 반대로 변경하려면 많은 로직을 손봐야 할 것 같아 고민을 많이 해봤는데요...! 🌱

공부 끝에 제네릭 열거형인 Result타입을 이용해 에러를 전달하기로 했고, 
saveButtonDidTap(_:)에서 `.success`와 `.failure`의 케이스 연관값으로 전달된 각각의 값·동작을 받아 실행하도록 구현하였어요!

이전과 달리 제네릭 조건(Error채택한 타입이어야 함) 탓에 
에러메시지(`error.localizedDescription`)가 아닌 (`error`)를 전달하게 되었으므로 함수 네이밍은 `detectInputError()`로 최종수정하였어요 !!!

### 질문 포인트
사실 .success는 이렇다 할 연관값이 없어서 고민이 되었어요. 
연관값을 아예 넣어주지 않는 방법을 알지 못해서 여러 후보를 거쳐 클로저를 넣어보기로 했어요. (3번↓)
1. `Result<Void, InputError>` ― `.success(())`
1. `Result<Bool, InputError>` ― `.success(true)`
1. `Result<() -> Void, InputError>` ― `.success({ self.dismiss(animated: true) })`

사실 Result형식을 맞추기 위해 불필요한 값이나 절차만 늘어난 듯 보이기도 해서 걱정이 돼요🥲🥲
이렇게 사용해도 괜찮은지, 더 좋은 방법이 있을지 조언 부탁드려요 코리🙏🏻

　
> 해당 커밋: 33a5699

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/ListProfileViewController.swift:
```diff
@@ -0,0 +1,72 @@
+        self.present(addProfileNav, animated: true)
+    }
+    
+    func updateProfile(name: String?, age: String?, tel: String?) {
+        guard let name, let age, let tel else { return }
```
파라미터 타입이 모두 옵셔널인데, 여기에서는 예외 처리를 거치고 실제로 데이터를 만들어야하는 단계이기 때문에, 이 전 단계인 AddProfileVC에서 옵셔널을 처리하면 좋을 것 같아요!

iyeahh replied:
두 컨트롤러의 역할을 고려할 때 말씀하신 대로 옵셔널을 처리해서 넘기는 게 더 알맞다고 생각했어요! 수정했습니다 😇
> 해당 커밋 https://github.com/tasty-code/ios-contact-manager-ui/pull/14/commits/5583b48c05a68dcf7677ede4afff52d90ccef36c

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+    @IBOutlet private weak var ageTextField: UITextField!
+    @IBOutlet private weak var telTextField: UITextField!
+    
+    weak var delegate: ListProfileViewController?
```
delegate를 weak로 잘 선언하셨습니다.
타입은 `ListProfileViewController?`으로 선언하는게 아니라 해당 프로토콜로 사용해주세요.

ueunli replied:
바로 수정했어요🥲 알려주셔서 감사해요..!!!

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+        return false
+    }
+    
+    private func insertHyphenInTel(_ tel: String) -> String {
```
조금 더 String을 반환할 것 같은 네이밍이면 좋겠네요!
또한 파라미터를 생략하는 와일드카드 패턴은 사용하지 않는게 호출 부분에서 사용할 때 더 깔끔합니다!
Swift 가이드 문서에 나와있는 방법을 사용하면 더 좋을 것 같아요 :)

iyeahh replied:
아직 네이밍이 익숙치 않아서 `String을 반활할 것 같은`의 느낌을 담는 것이 어렵더라구요...!
`hyphenate(phoneNumber:)` 으로 변경하였는데 더 좋은 네이밍이 있다면 추천 부탁드릴게요 🥹

> 해당 커밋 [dc61abb](https://github.com/tasty-code/ios-contact-manager-ui/pull/14/commits/dc61abbd6451b3a02997082332eaa4d2486c5a25)

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,131 @@
+    private func insertHyphenInTel(_ tel: String) -> String {
+        let tel = tel.replacingOccurrences(of: "-", with: "")
+        switch tel.count {
+        case ...3: return PhoneNumberRegularExpressions.oneToThree.transform(tel)
```
여유가 된다면 반복되는 transform 함수를 정리하면 더 좋을 것 같네요 😊

ueunli replied:
`PhoneNumberRegularExpressions`열거형에 위의 `tel.count`에 따라 케이스 인스턴스를 생성해주는 이니셜라이저를 새로 정의하여 해결했어요!
해당 부분에서는 `생성된 인스턴스`.transform(`tel`) 1줄로 통합하여 호출할 수 있게 되었는데, 구조적으로 괜찮은 방법인지는 확신이 서지 않아요... 확인 부탁드려요 !!
> 해당 커밋: 8f45cff

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/ListProfileViewController.swift:
```diff
@@ -0,0 +1,72 @@
+
+import UIKit
+
+protocol ListProfileViewControllerDelegate {
```
Delegate의 정의가 무엇인가요?
ListProfileViewControllerDelegate는 어떤 역할인건가요?
(어떤 객체의 어떤 역할을 하는지 이야기해주면 좋을 것 같아요!)

ueunli replied:
Delegate의 정의는 _**위임하다**_ 예요. 
개인적으로 Delegate Pattern이란 _**책임자-대리자 패턴**_ 이라고 이해하고 있어요

`ListProfileViewController`는 새로운 뷰를 올릴 때 본인을 그 뷰(컨트롤러)의 대리자로 지정하여 함께 보내지고, 
`ListProfileViewControllerDelegate`(ProfileDelegate)는 `AddProfileViewController` 내에 머물며(= delegate변수에 할당) 대리자로서 요구받은 동작(프로토콜 필수구현 메서드)을 적절한 시점에 수행해요.

- 저희 코드에서 그 동작은 `AddProfileViewController`에서 Save버튼이 눌렀을 때 호출되어, 
- (검증이 완료된) 새로운 이름·나이·연락처 정보를 조합하여 프로필을 생성해요(= 요구받은 동작).
- `AddProfileViewController`가 닫히면 그 데이터(새 프로필)를 받아 필요한 작업(`profiles`에 새 프로필을 등록)을 이어서 수행하게 돼요!

위에 남겨주신 피드백 중 델리게이트 메서드의 매개변수들을 `String?`이 아닌 `String`으로 받아오라고 조언해주신 것도 비슷한 맥락에서 '대리자'가 위임받은 역할에 충실하기 위해서였다고 이해하였고
덕분에 델리게이트 패턴에 대한 이해가 더욱 깊어질 수 있었어요 !! 감사합니다🙇🏻‍♀️

　
> +) 프로토콜의 역할을 고려하여 네이밍을 `ListProfileViewControllerDelegate` → `ProfileDelegate`로 변경하였습니다 !! (해당 커밋: 6e412db)

corykim0829 commented on ios-cantact-manager/ios-cantact-manager/ContactManageSystem.swift:
```diff
@@ -76,24 +55,8 @@ struct ContactManageSystem {
-        } catch {
-            OutputManager.print(text: .invalidInput)
-        }
+    mutating func addProfile(of profile: Profile) {
```
파라미터에 of라고 하면 문법상 조금 어색하네요..!
스위프트 가이드 문서 참고해서 수정하면 좋을 것 같아요!

iyeahh replied:
of를 삭제하고 `addProfile(_:)`로 변경했는데 혹시 `add(profile:)`이 더 나았을지 궁금합니다...!!!

> 해당 커밋 [dc61abb](https://github.com/tasty-code/ios-contact-manager-ui/pull/14/commits/dc61abbd6451b3a02997082332eaa4d2486c5a25)

corykim0829 reviewed (changes requested):
두분 수고 많으셨습니다!👏👏👏👏

UI 구현은 전반적으로 깔끔하게 잘 하셨습니다 👍

같이 더 고민해보고 싶은 부분이 있다면 편하게 말씀해주세요!

iyeahh committed:
<refactor> 피드백 반영하여 전반적인 코드 수정

- 불필요한 중복 코드 및 오타 수정
- present alert 위치 변경
- 특정 textField 대조 방법 변경 (.phonePad -> 해당 인스턴스)

ueunli committed:
<refactor> ListProfileViewControllerDelegate의 메서드 수정

- 파라미터 타입을 `String?` → `String`으로 변경

ueunli committed:
<refactor> 에러핸들링 및 얼럿 로직 수정

- `detectInputError`의 반환형을 String? → Result타입으로 변경하여 에러 여부에 따른 로직을 기존과 반대로 수정
- `generateAlertMessage` → `detectInputError` 함수명 변경

ueunli committed:
<refactor> 반복되는 `transform(_:)`메서드 호출을 간소화

- `PhoneNumberRegularExpressions`열거형에 생성자 추가

iyeahh committed:
<chore> 네이밍 수정

- addProfile(of:) -> addProfile(_:)
- insertHyphenInTel(_:) -> hyphenate(phoneNumber:)

iyeahh committed:
<chore> 델리게이트 프로토콜 네이밍 변경

- ListProfileViewController -> Profile

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/ListProfileViewController.swift:
```diff
@@ -0,0 +1,71 @@
+
+import UIKit
+
+protocol ProfileDelegate: AnyObject {
```
ProfileDelegate라고 하면 조금 추상적이네요.
이 프로토콜을 통해서 어떤 작업을 하는지 더 명확하게 나타나면 좋을 것 같아요.

또한 이 프로토콜의 구현 위치는 ListProfileVC가 적합해 보이지 않네요! 이 부분도 고민해보고 수정하면 좋을 것 같아요 😊

iyeahh replied:
Delegate 네이밍과 작성 위치에 대해 잘 모르고 있다가 질문을 받고
`Delegate naming conventions` 에서 `delegate 프로토콜의 이름은 클래스 이름으로 시작해야 한다` 를 발견하여
`AddProfileViewControllerDelegate`로 네이밍 변경하였고,
프로토콜 선언 위치는 해당 클래스 위에 작성해 두었습니다!

이것이 적절한 것인지 확인 부탁드리고 싶어요 🤓

corykim0829 replied:
넵 정석적인 방법입니다 👍

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        return false
+    }
+    
+    private func hyphenate(phoneNumber: String) -> String {
```
함수명은 좋습니다 👍👍
phoneNumber, tel이 혼용되는 부분은 통일하면 더 깔끔할 것 같아요!

iyeahh replied:
앗! 통일을 위해 phoneNumber -> tel로 변경했습니다 👍🏻

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        return regex.transform(tel)
+    }
+    
+    private enum PhoneNumberRegularExpressions: String {
```
enum에는 복수형을 쓰기보다는 단수형을 쓰는게 좋습니다 :)

iyeahh replied:
아하 또 하나 배웠습니다 😇 변경했어요 :)

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        case ten = "([0-9]{3})([0-9]{3})([0-9]{4})"
+        case elevenOrMore = "([0-9]{3})([0-9]{4})([0-9]{4,})"
+        
+        private var regexTemplate: String {
```
regex 라는 enum 안이기 때문에 template라고 표현하면 더 깔끔할 것 같아요!

iyeahh replied:
중복을 피하기 위해 regexTemplate - > template으로 변경했어요 🙋🏻‍♀️

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,139 @@
+            }
+        }
+        
+        func transform(_ phoneNumber: String) -> String {
```
transform은 UI요소를 변화시키는 의미로 많이 사용되어서 다른 용어로 바꾸면 더 가독성이 좋을 것 같아요! 참고만하세요 :)

ueunli replied:
양식을 적용한다는 의미로 `transform` → `format`으로 변경, 언더바 삭제하였습니다!
이곳의 'phoneNumber'도 'tel'로 통일할까 하다가, 속한 열거형 네이밍과 맞추는 편이 좋을 것 같아 유지했어요 :D

+) 관습적인 의미를 갖는 용어들도 좀 더 찾아보고 알아두겠습니다 !! 알려주셔서 감사해요🙏🏻

corykim0829 commented on ios-cantact-manager/ContactManagerUI/Controllers/AddProfileViewController.swift:
```diff
@@ -0,0 +1,139 @@
+            return result
+        }
+        
+        init(phoneNumber: String) {
```
initializer는 case 바로 아래로 이동하면 좋을 것 같고, private 함수보다는 public 함수가 위로 가야 전체적으로 더 보기 좋을 것 같아요!

iyeahh replied:
기억해두었다가 앞으로도 적용하도록 하겠습니다! 🤔 변경했어요 :)

corykim0829 commented on ios-cantact-manager/ios-cantact-manager/ContactManageSystem.swift:
```diff
@@ -76,24 +55,8 @@ struct ContactManageSystem {
-        } catch {
-            OutputManager.print(text: .invalidInput)
-        }
+    mutating func addProfile(_ profile: Profile) {
```
add(profile: Profile)이 스위프트 가이드 문서를 더 준수하는 방법입니다! 참고하세요 :)

iyeahh replied:
네이밍은 여전히 어렵네요 ㅠㅠ add(profile: Profile)로 변경했습니다 🥹

corykim0829 reviewed (changes requested):
수고 많으셨습니다 👏👏👏

고민을 통해서 더 좋은 코드로 개선해나가는 과정이 저도 잘 느껴져서 좋네요 😊

제가 남겨드린 커멘트 확인해주시고 더 개선해나가봅시다!

+추가로 보너스 미션은 진행하실 예정인지 궁금합니다! 만약 하신다면, 두개 중 하나를 선택해서 시작하면 좋을 것 같습니다! 🚀

ueunli committed:
<chore> 네이밍 및 메서드 배치 수정

iyeahh committed:
<chore> 프로토콜 위치 및 네이밍 변경

corykim0829 reviewed (approved):
수고 많으셨습니다 😊
